C++ SampleMutUninit payload operator: error: taking address of rvalue

## Required information

**Operating system:**
Ubuntu 24.04.3 LTS
`Linux GLDLBAE00056713 6.6.87.2-microsoft-standard-WSL2 #1 SMP PREEMPT_DYNAMIC Thu Jun  5 18:30:46 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux`

**Rust version:**
`rustc 1.90.0 (1159e78c4 2025-09-14)`

**Cargo version:**
`cargo 1.90.0 (840b83a10 2025-07-30)`

**iceoryx2 version:**
`v0.7.0`

**clang version:**
`Ubuntu clang version 20.1.8 (++20250804090239+87f0227cb601-1~exp1~20250804210352.139)`

**gcc version:**
`gcc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0`

**c++ std:**
`c++20`

**Detailed log output:**
Clang does not complain. GCC does complain, though it can be supressed with `-fpermissive`.
```
/home/blake/.conan2/p/b/iceor64e5d76a8f489/p/include/iceoryx2/v0.7.0/iox2/sample_mut_uninit.hpp:124:24: error: taking address of rvalue [-fpermissive]
  124 |     return &payload_mut();
      |             ~~~~~~~~~~~^~
```

**Observed result or behaviour:**
At optimization level `-O2` or above, this causes runtime crashes under clang. GCC requires `-fpermissive` to allow this to compile, when doing so it will crash on (all?) optimization levels.

**Expected result or behaviour:**
Not crashing.

**Conditions where it occurred / Performed steps:**
Create a sample and try to access it's payload, the program will segfault. Example:
```
auto sample = publisher.loan_slice_uninit( messageWords * sizeof( capnp::word ) ).expect( "accquire sample" );
SPDLOG_INFO("Payload bytes: {}", sample->number_of_bytes());
```

By changing the above to explicitly retrieve the payload from the sample, the crash is fixed and the code works as expected.
```
auto sample = publisher.loan_slice_uninit( messageWords * sizeof( capnp::word ) ).expect( "accquire sample" );
auto payload = sample.payload_mut();
SPDLOG_INFO("Payload bytes: {}", payload.number_of_bytes());
```
I would assume this same issue might appear in other places as well, but I haven't checked myself.

--- a/doc/release-notes/iceoryx2-unreleased.md
+++ b/doc/release-notes/iceoryx2-unreleased.md
@@ -71,7 +71,7 @@
 
 ### API Breaking Changes
 
-1. Replaced the `FixedSizeVec` with the `StaticVec`
+1. **Rust:** Replaced the `FixedSizeVec` with the `StaticVec`
 
    ```rust
    // old
@@ -85,7 +85,7 @@
    let my_vec = StaticVec::<MyType, VEC_CAPACITY>::new();
    ```
 
-2. Replaced `Vec` with the `PolymorphicVec`
+2. **Rust:** Replaced `Vec` with the `PolymorphicVec`
 
     ```rust
    // old
@@ -100,7 +100,7 @@
    let my_vec = PolymorphicVec::<MyType>::new(my_stateful_allocator, vec_capacity)?;
     ```
 
-3. Replaced the `FixedSizeByteString` with the `StaticString`
+3. **Rust:** Replaced the `FixedSizeByteString` with the `StaticString`
 
    ```rust
    // old
@@ -112,4 +112,33 @@
    use iceoryx2_bb_container::string::*;
    const CAPACITY: usize = 1234;
    let my_str = StaticString::<CAPACITY>::new();
+   ```
+
+4. **C++:** Remove `operator*` and `operator->` from `ActiveRequest`,
+   `PendingResponse`, `RequestMut`, `RequestMutUninit`, `Response`,
+   `ResponseMut`, `Sample`, `SampleMut`, `SampleMutUninit` since it lead
+   to easy confusions and bugs when used in combination with `optional` or
+   `expected`. See `sample.has_value()` and `sample->has_value()` that work
+   on different objects.
+
+   ```cxx
+   // old
+   auto sample = publisher.loan().expect("");
+   sample->some_member = 123;
+
+   // new
+   auto sample = publisher.loan().expect("");
+   sample.payload_mut().some_member = 123;
+   ```
+
+   ```cxx
+   // old
+   auto sample = publisher.loan().expect("");
+   *sample = 123;
+   std::cout << *sample << std::endl;
+
+   // new
+   auto sample = publisher.loan().expect("");
+   sample.payload_mut() = 123;
+   std::cout << sample.payload() << std::endl;
    ```

--- a/examples/cxx/health_monitoring/src/subscriber.cpp
+++ b/examples/cxx/health_monitoring/src/subscriber.cpp
@@ -109,7 +109,7 @@ void handle_incoming_events(Listener<ServiceType::Ipc>& listener,
                 std::cout << service_name.to_string().c_str() << ": publisher disconnected!" << std::endl;
             } else if (event_id == iox::into<EventId>(PubSubEvent::SentSample)) {
                 subscriber.receive().expect("").and_then([&](auto& sample) {
-                    std::cout << service_name.to_string().c_str() << ": Received sample " << *sample << " ..."
+                    std::cout << service_name.to_string().c_str() << ": Received sample " << sample.payload() << " ..."
                               << std::endl;
                 });
             }

--- a/examples/cxx/request_response/src/server.cpp
+++ b/examples/cxx/request_response/src/server.cpp
@@ -47,7 +47,7 @@ auto main() -> int {
                     auto response = active_request->loan_uninit().expect("loan successful");
                     auto initialized_response = response.write_payload(
                         TransmissionData { counter * (iter + 1), counter + iter, counter * 0.1234 }); // NOLINT
-                    std::cout << "send response: " << *initialized_response << std::endl;
+                    std::cout << "send response: " << initialized_response.payload() << std::endl;
                     send(std::move(initialized_response)).expect("send successful");
                 }
             } else {

--- a/iceoryx2-cxx/include/iox2/active_request.hpp
+++ b/iceoryx2-cxx/include/iox2/active_request.hpp
@@ -40,9 +40,6 @@ class ActiveRequest {
     ActiveRequest(const ActiveRequest&) = delete;
     auto operator=(const ActiveRequest&) noexcept -> ActiveRequest& = delete;
 
-    auto operator*() const -> const RequestPayload&;
-    auto operator->() const -> const RequestPayload*;
-
     /// Loans uninitialized memory for a [`ResponseMutUninit`] where the user can write its payload to.
     template <typename T = ResponsePayload, typename = std::enable_if_t<!iox::IsSlice<T>::VALUE, void>>
     auto loan_uninit() -> iox::expected<ResponseMutUninit<Service, ResponsePayload, ResponseUserHeader>, LoanError>;
@@ -146,28 +143,6 @@ template <ServiceType Service,
 inline ActiveRequest<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::
     ~ActiveRequest() noexcept {
     drop();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-ActiveRequest<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator*() const
-    -> const RequestPayload& {
-    return payload();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-ActiveRequest<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator->() const
-    -> const RequestPayload* {
-    return &payload();
 }
 
 template <ServiceType Service,

--- a/iceoryx2-cxx/include/iox2/pending_response.hpp
+++ b/iceoryx2-cxx/include/iox2/pending_response.hpp
@@ -47,9 +47,6 @@ class PendingResponse {
     PendingResponse(const PendingResponse&) = delete;
     auto operator=(const PendingResponse&) -> PendingResponse& = delete;
 
-    auto operator*() const -> const RequestPayload&;
-    auto operator->() const -> const RequestPayload*;
-
     /// Receives a [`Response`] from one of the [`Server`]s that
     /// received the [`RequestMut`].
     auto receive()
@@ -129,47 +126,9 @@ template <ServiceType Service,
           typename RequestUserHeader,
           typename ResponsePayload,
           typename ResponseUserHeader>
-inline auto PendingResponse<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator=(
-    PendingResponse&& rhs) noexcept -> PendingResponse& {
-    if (this != &rhs) {
-        drop();
-        m_handle = std::move(rhs.m_handle);
-        rhs.m_handle = nullptr;
-    }
-
-    return *this;
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
 inline PendingResponse<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::
     ~PendingResponse() noexcept {
     drop();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-PendingResponse<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator*() const
-    -> const RequestPayload& {
-    return payload();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-PendingResponse<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator->() const
-    -> const RequestPayload* {
-    return &payload();
 }
 
 template <ServiceType Service,

--- a/iceoryx2-cxx/include/iox2/pending_response.hpp
+++ b/iceoryx2-cxx/include/iox2/pending_response.hpp
@@ -126,6 +126,22 @@ template <ServiceType Service,
           typename RequestUserHeader,
           typename ResponsePayload,
           typename ResponseUserHeader>
+inline auto PendingResponse<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator=(
+    PendingResponse&& rhs) noexcept -> PendingResponse& {
+    if (this != &rhs) {
+        drop();
+        m_handle = std::move(rhs.m_handle);
+        rhs.m_handle = nullptr;
+    }
+
+    return *this;
+}
+
+template <ServiceType Service,
+          typename RequestPayload,
+          typename RequestUserHeader,
+          typename ResponsePayload,
+          typename ResponseUserHeader>
 inline PendingResponse<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::
     ~PendingResponse() noexcept {
     drop();

--- a/iceoryx2-cxx/include/iox2/request_mut.hpp
+++ b/iceoryx2-cxx/include/iox2/request_mut.hpp
@@ -43,18 +43,6 @@ class RequestMut {
     RequestMut(const RequestMut&) = delete;
     auto operator=(const RequestMut&) -> RequestMut& = delete;
 
-    /// Returns a const reference to the request payload
-    auto operator*() const -> const RequestPayload&;
-
-    /// Returns a reference to the request payload
-    auto operator*() -> RequestPayload&;
-
-    /// Returns a const pointer to the request payload
-    auto operator->() const -> const RequestPayload*;
-
-    /// Returns a pointer to the request payload
-    auto operator->() -> RequestPayload*;
-
     /// Returns a reference to the iceoryx2 internal [`RequestHeader`]
     auto header() const -> RequestHeader;
 
@@ -149,48 +137,6 @@ template <ServiceType Service,
 inline RequestMut<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::
     ~RequestMut() noexcept {
     drop();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-RequestMut<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator*() const
-    -> const RequestPayload& {
-    return payload();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto RequestMut<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator*()
-    -> RequestPayload& {
-    return payload_mut();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-RequestMut<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator->() const
-    -> const RequestPayload* {
-    return &payload();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto RequestMut<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator->()
-    -> RequestPayload* {
-    return &payload_mut();
 }
 
 template <ServiceType Service,

--- a/iceoryx2-cxx/include/iox2/request_mut_uninit.hpp
+++ b/iceoryx2-cxx/include/iox2/request_mut_uninit.hpp
@@ -39,18 +39,6 @@ class RequestMutUninit {
     RequestMutUninit(const RequestMutUninit&) = delete;
     auto operator=(const RequestMutUninit&) -> RequestMutUninit& = delete;
 
-    /// Returns a const reference to the request payload
-    auto operator*() const -> const RequestPayload&;
-
-    /// Returns a reference to the request payload
-    auto operator*() -> RequestPayload&;
-
-    /// Returns a const pointer to the request payload
-    auto operator->() const -> const RequestPayload*;
-
-    /// Returns a pointer to the request payload
-    auto operator->() -> RequestPayload*;
-
     /// Returns a reference to the iceoryx2 internal [`RequestHeader`]
     auto header() const -> RequestHeader;
 
@@ -113,50 +101,6 @@ class RequestMutUninit {
 
     RequestMut<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader> m_request;
 };
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-RequestMutUninit<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator*() const
-    -> const RequestPayload& {
-    return payload();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-RequestMutUninit<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator*()
-    -> RequestPayload& {
-    return payload_mut();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-RequestMutUninit<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator->() const
-    -> const RequestPayload* {
-    return &payload();
-}
-
-template <ServiceType Service,
-          typename RequestPayload,
-          typename RequestUserHeader,
-          typename ResponsePayload,
-          typename ResponseUserHeader>
-inline auto
-RequestMutUninit<Service, RequestPayload, RequestUserHeader, ResponsePayload, ResponseUserHeader>::operator->()
-    -> RequestPayload* {
-    return &payload_mut();
-}
 
 template <ServiceType Service,
           typename RequestPayload,

--- a/iceoryx2-cxx/include/iox2/response.hpp
+++ b/iceoryx2-cxx/include/iox2/response.hpp
@@ -33,9 +33,6 @@ class Response {
     Response(const Response&) noexcept = delete;
     auto operator=(const Response&) noexcept -> Response& = delete;
 
-    auto operator*() const -> const ResponsePayload&;
-    auto operator->() const -> const ResponsePayload*;
-
     /// Returns a reference to the [`ResponseHeader`].
     auto header() const -> ResponseHeader;
 
@@ -86,16 +83,6 @@ inline auto Response<Service, ResponsePayload, ResponseUserHeader>::operator=(Re
 template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
 inline Response<Service, ResponsePayload, ResponseUserHeader>::~Response() noexcept {
     drop();
-}
-
-template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
-inline auto Response<Service, ResponsePayload, ResponseUserHeader>::operator*() const -> const ResponsePayload& {
-    return payload();
-}
-
-template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
-inline auto Response<Service, ResponsePayload, ResponseUserHeader>::operator->() const -> const ResponsePayload* {
-    return &payload();
 }
 
 template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>

--- a/iceoryx2-cxx/include/iox2/response_mut.hpp
+++ b/iceoryx2-cxx/include/iox2/response_mut.hpp
@@ -43,11 +43,6 @@ class ResponseMut {
     ResponseMut(const ResponseMut&) = delete;
     auto operator=(const ResponseMut&) -> ResponseMut& = delete;
 
-    auto operator*() const -> const ResponsePayload&;
-    auto operator*() -> ResponsePayload&;
-    auto operator->() const -> const ResponsePayload*;
-    auto operator->() -> ResponsePayload*;
-
     /// Returns a reference to the [`ResponseHeader`].
     auto header() const -> ResponseHeader;
 
@@ -124,26 +119,6 @@ inline auto ResponseMut<Service, ResponsePayload, ResponseUserHeader>::operator=
 template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
 inline ResponseMut<Service, ResponsePayload, ResponseUserHeader>::~ResponseMut() noexcept {
     drop();
-}
-
-template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
-inline auto ResponseMut<Service, ResponsePayload, ResponseUserHeader>::operator*() const -> const ResponsePayload& {
-    return payload();
-}
-
-template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
-inline auto ResponseMut<Service, ResponsePayload, ResponseUserHeader>::operator*() -> ResponsePayload& {
-    return payload_mut();
-}
-
-template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
-inline auto ResponseMut<Service, ResponsePayload, ResponseUserHeader>::operator->() const -> const ResponsePayload* {
-    return &payload();
-}
-
-template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>
-inline auto ResponseMut<Service, ResponsePayload, ResponseUserHeader>::operator->() -> ResponsePayload* {
-    return &payload_mut();
 }
 
 template <ServiceType Service, typename ResponsePayload, typename ResponseUserHeader>

--- a/iceoryx2-cxx/include/iox2/sample.hpp
+++ b/iceoryx2-cxx/include/iox2/sample.hpp
@@ -47,12 +47,6 @@ class Sample {
     auto operator=(const Sample&) -> Sample& = delete;
 
     /// Returns a reference to the payload of the [`Sample`]
-    auto operator*() const -> const Payload&;
-
-    /// Returns a pointer to the payload of the [`Sample`]
-    auto operator->() const -> const Payload*;
-
-    /// Returns a reference to the payload of the [`Sample`]
     template <typename T = Payload, typename = std::enable_if_t<!iox::IsSlice<T>::VALUE, void>>
     auto payload() const -> const ValueType&;
 
@@ -117,16 +111,6 @@ inline auto Sample<S, Payload, UserHeader>::operator=(Sample&& rhs) noexcept -> 
 template <ServiceType S, typename Payload, typename UserHeader>
 inline Sample<S, Payload, UserHeader>::~Sample() {
     drop();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto Sample<S, Payload, UserHeader>::operator*() const -> const Payload& {
-    return payload();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto Sample<S, Payload, UserHeader>::operator->() const -> const Payload* {
-    return &payload();
 }
 
 template <ServiceType S, typename Payload, typename UserHeader>

--- a/iceoryx2-cxx/include/iox2/sample_mut.hpp
+++ b/iceoryx2-cxx/include/iox2/sample_mut.hpp
@@ -55,18 +55,6 @@ class SampleMut {
     SampleMut(const SampleMut&) = delete;
     auto operator=(const SampleMut&) -> SampleMut& = delete;
 
-    /// Returns a const reference to the payload of the [`Sample`]
-    auto operator*() const -> const Payload&;
-
-    /// Returns a reference to the payload of the [`Sample`]
-    auto operator*() -> Payload&;
-
-    /// Returns a const pointer to the payload of the [`Sample`]
-    auto operator->() const -> const Payload*;
-
-    /// Returns a pointer to the payload of the [`Sample`]
-    auto operator->() -> Payload*;
-
     /// Returns a reference to the [`Header`] of the [`Sample`].
     auto header() const -> HeaderPublishSubscribe;
 
@@ -146,26 +134,6 @@ inline auto SampleMut<S, Payload, UserHeader>::operator=(SampleMut&& rhs) noexce
 template <ServiceType S, typename Payload, typename UserHeader>
 inline SampleMut<S, Payload, UserHeader>::~SampleMut() noexcept {
     drop();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMut<S, Payload, UserHeader>::operator*() const -> const Payload& {
-    return payload();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMut<S, Payload, UserHeader>::operator*() -> Payload& {
-    return payload_mut();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMut<S, Payload, UserHeader>::operator->() const -> const Payload* {
-    return &payload();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMut<S, Payload, UserHeader>::operator->() -> Payload* {
-    return &payload_mut();
 }
 
 template <ServiceType S, typename Payload, typename UserHeader>

--- a/iceoryx2-cxx/include/iox2/sample_mut_uninit.hpp
+++ b/iceoryx2-cxx/include/iox2/sample_mut_uninit.hpp
@@ -33,18 +33,6 @@ class SampleMutUninit {
     SampleMutUninit(const SampleMutUninit&) = delete;
     auto operator=(const SampleMutUninit&) -> SampleMutUninit& = delete;
 
-    /// Returns a const reference to the payload of the [`Sample`]
-    auto operator*() const -> const Payload&;
-
-    /// Returns a reference to the payload of the [`Sample`]
-    auto operator*() -> Payload&;
-
-    /// Returns a const pointer to the payload of the [`Sample`]
-    auto operator->() const -> const Payload*;
-
-    /// Returns a pointer to the payload of the [`Sample`]
-    auto operator->() -> Payload*;
-
     /// Returns a reference to the [`Header`] of the [`Sample`].
     auto header() const -> HeaderPublishSubscribe;
 
@@ -102,26 +90,6 @@ class SampleMutUninit {
 template <ServiceType S, typename Payload, typename UserHeader>
 inline auto assume_init(SampleMutUninit<S, Payload, UserHeader>&& self) -> SampleMut<S, Payload, UserHeader> {
     return std::move(self.m_sample);
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMutUninit<S, Payload, UserHeader>::operator*() const -> const Payload& {
-    return payload();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMutUninit<S, Payload, UserHeader>::operator*() -> Payload& {
-    return payload_mut();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMutUninit<S, Payload, UserHeader>::operator->() const -> const Payload* {
-    return &payload();
-}
-
-template <ServiceType S, typename Payload, typename UserHeader>
-inline auto SampleMutUninit<S, Payload, UserHeader>::operator->() -> Payload* {
-    return &payload_mut();
 }
 
 template <ServiceType S, typename Payload, typename UserHeader>

--- a/iceoryx2-cxx/tests/src/service_publish_subscribe_tests.cpp
+++ b/iceoryx2-cxx/tests/src/service_publish_subscribe_tests.cpp
@@ -242,7 +242,7 @@ TYPED_TEST(ServicePublishSubscribeTest, send_copy_receive_works) {
     auto sample = sut_subscriber.receive().expect("");
 
     ASSERT_TRUE(sample.has_value());
-    ASSERT_THAT(**sample, Eq(payload));
+    ASSERT_THAT(sample->payload(), Eq(payload));
 }
 
 TYPED_TEST(ServicePublishSubscribeTest, loan_send_receive_works) {
@@ -258,12 +258,12 @@ TYPED_TEST(ServicePublishSubscribeTest, loan_send_receive_works) {
 
     auto sample = sut_publisher.loan().expect("");
     const uint64_t payload = 781891729871;
-    *sample = payload;
+    sample.payload_mut() = payload;
     send(std::move(sample)).expect("");
     auto recv_sample = sut_subscriber.receive().expect("");
 
     ASSERT_TRUE(recv_sample.has_value());
-    ASSERT_THAT(**recv_sample, Eq(payload));
+    ASSERT_THAT(recv_sample->payload(), Eq(payload));
 }
 
 TYPED_TEST(ServicePublishSubscribeTest, loan_uninit_send_receive_works) {
@@ -284,7 +284,7 @@ TYPED_TEST(ServicePublishSubscribeTest, loan_uninit_send_receive_works) {
     auto recv_sample = sut_subscriber.receive().expect("");
 
     ASSERT_TRUE(recv_sample.has_value());
-    ASSERT_THAT(**recv_sample, Eq(payload));
+    ASSERT_THAT(recv_sample->payload(), Eq(payload));
 }
 
 struct DummyData {
@@ -564,7 +564,7 @@ TYPED_TEST(ServicePublishSubscribeTest, update_connections_delivers_history) {
     sample = sut_subscriber.receive().expect("");
 
     ASSERT_TRUE(sample.has_value());
-    ASSERT_THAT(**sample, Eq(payload));
+    ASSERT_THAT(sample->payload(), Eq(payload));
 }
 
 TYPED_TEST(ServicePublishSubscribeTest, setting_service_properties_works) {
@@ -732,7 +732,7 @@ TYPED_TEST(ServicePublishSubscribeTest, send_receive_with_user_header_works) {
 
     auto sample = sut_publisher.loan().expect("");
     const uint64_t payload = 781891729871;
-    *sample = payload;
+    sample.payload_mut() = payload;
     for (uint64_t idx = 0; idx < TestHeader::CAPACITY; ++idx) {
         sample.user_header_mut().value.at(idx) = 4 * idx + 3;
     }
@@ -740,7 +740,7 @@ TYPED_TEST(ServicePublishSubscribeTest, send_receive_with_user_header_works) {
     auto recv_sample = sut_subscriber.receive().expect("");
 
     ASSERT_TRUE(recv_sample.has_value());
-    ASSERT_THAT(**recv_sample, Eq(payload));
+    ASSERT_THAT(recv_sample->payload(), Eq(payload));
 
     for (uint64_t idx = 0; idx < TestHeader::CAPACITY; ++idx) {
         ASSERT_THAT(recv_sample->user_header().value.at(idx), Eq((4 * idx) + 3));

--- a/iceoryx2-cxx/tests/src/service_request_response_tests.cpp
+++ b/iceoryx2-cxx/tests/src/service_request_response_tests.cpp
@@ -359,22 +359,19 @@ TYPED_TEST(ServiceRequestResponseTest, loan_send_receive_works) {
     EXPECT_THAT(request.payload().p, Eq(3));
 
     auto pending_response = send(std::move(request)).expect("");
-    EXPECT_THAT(pending_response->p, Eq(3));
-    EXPECT_THAT((*pending_response).p, Eq(3));
+    EXPECT_THAT(pending_response.payload().p, Eq(3));
 
     auto active_request = sut_server.receive().expect("");
     ASSERT_TRUE(active_request.has_value());
-    EXPECT_THAT(active_request.value()->p, Eq(3));
-    EXPECT_THAT((*active_request.value()).p, Eq(3));
+    EXPECT_THAT(active_request.value().payload().p, Eq(3));
 
     auto response = active_request->loan().expect("");
-    response->p = 0;
+    response.payload_mut().p = 0;
     send(std::move(response)).expect("");
 
     auto received_response = pending_response.receive().expect("");
     ASSERT_TRUE(received_response.has_value());
-    EXPECT_THAT(received_response.value()->p, Eq(0));
-    EXPECT_THAT((*received_response.value()).p, Eq(0));
+    EXPECT_THAT(received_response.value().payload().p, Eq(0));
 }
 
 TYPED_TEST(ServiceRequestResponseTest, loan_request_default_constructs_request_header) {
@@ -949,7 +946,7 @@ TYPED_TEST(ServiceRequestResponseTest, send_receive_with_user_header_works) {
     EXPECT_THAT(active_request->user_header(), Eq(request_user_header));
 
     auto response = active_request->loan().expect("");
-    *response = 2;
+    response.payload_mut() = 2;
     response.user_header_mut() = response_user_header;
     send(std::move(response)).expect("");
 

